Update GAT tutorial on mapping jobs to destinations using TPV (#3983)
galaxyproject/training-material: topics/admin/tutorials/job-destinations/tutorial.md

  - natefoo
  - bgruening
  - hexylena
  - nuwang
  - cat-bro

>   - afgane

Reviewer: If you like can use our fancy new contributions syntax to more accurately record contributions to a training material, declaring which people should be recorded for authorship or just editing. (I think `natefoo` and `bgruening` can claim `authorship`, I'm probably an `editing` since I mostly did the GTN conversion of this tutorial)

```
contributions:
  authorship:
    - natefoo
    - bgruening
  editing:
    - hexylena
    - hexylena
  funding: []
  testing: []
```

see the [full gtn tutorial schema documentation](https://training.galaxyproject.org/training-material/topics/contributing/tutorials/schemas/tutorial.html) for details on the individual roles.

  - bgruening

>   - hexylena

Reviewer 2: ```suggestion
  - hexylena
  - nuwang
```

{% snippet topics/admin/faqs/git-gat-path.md tutorial="job-destinations" %}

# Mapping jobs to destinations

In order to run jobs in Galaxy, you need to assign them to a resource manager that can handle the task. This involves specifying the appropriate amount of memory and CPU cores. For production installations, the jobs
must be routed to a resource manager like SLURM, HTCondor, or Pulsar. Some tools may need specific resources such as GPUs or multi-core machines to work efficiently.

Sometimes, your available resources are spread out across multiple locations and resource managers. In such cases, you need a way to route your jobs to the appropriate location. Galaxy offers several methods for

> routing jobs, ranging from simple static mappings to custom Python functions called dynamic job destinations.

Reviewer 2: ```suggestion
routing jobs, ranging from simple static mappings to custom Python functions via [dynamic job destinations](https://docs.galaxyproject.org/en/latest/admin/jobs.html#dynamic-destination-mapping).
```

> Recently, the Galaxy project has introduced a library named Total-Perspective-Vortex (TPV) to simplify this process. TPV provides a user-friendly YAML configuration that works for most scenarios.

Reviewer 2: ```suggestion
Recently, the Galaxy project has introduced a library named [Total-Perspective-Vortex (TPV)](https://total-perspective-vortex.readthedocs.io/) to simplify this process. TPV provides a admin-friendly YAML configuration that works for most scenarios.
```

> For more complex cases, TPV allows you to embed Python code into the YAML file. Additionally, TPV shares a global database of resource requirements, so admins don't have to figure out the requirements for each tool

Reviewer 2: ```suggestion
For more complex cases, TPV also allows you to embed Python code into the configuration YAML file and implement fine-grained control over jobs. 

Lastly, TPV offers a shared global database of default resource requirements (more below). By leveraging this database, admins don't have to figure out the requirements for each tool
```

separately.

## Writing a testing tool

> To check that resources are being allocated appropriately, We don't want to overload our training VMs trying to run real tools, so to demonstrate how to map a multicore tool to a multicore destination using TPV,

Reviewer 2: ```suggestion
To demonstrate a real-life scenario and TPV's role, let's plan on setting up a configuration so that a VM designated for training jobs doesn't run and real jobs and hence doesn't get overloaded. To start, we'll create a fake tool that we'll use in our configuration.
```

> we'll create a fake tool.

Reviewer 2: ```suggestion

```

must be routed to a resource manager like SLURM, HTCondor, or Pulsar. Some tools may need specific resources such as GPUs or multi-core machines to work efficiently.

Sometimes, your available resources are spread out across multiple locations and resource managers. In such cases, you need a way to route your jobs to the appropriate location. Galaxy offers several methods for
routing jobs, ranging from simple static mappings to custom Python functions via [dynamic job destinations](https://docs.galaxyproject.org/en/latest/admin/jobs.html#dynamic-destination-mapping).

Recently, the Galaxy project has introduced a library named [Total-Perspective-Vortex (TPV)](https://total-perspective-vortex.readthedocs.io/) to simplify this process. TPV provides a admin-friendly YAML configuration that works for most scenarios.
For more complex cases, TPV also allows you to embed Python code into the configuration YAML file and implement fine-grained control over jobs. 

Lastly, TPV offers a shared global database of default resource requirements (more below). By leveraging this database, admins don't have to figure out the requirements for each tool
separately.

## Writing a testing tool

> To demonstrate a real-life scenario and TPV's role in it, let's plan on setting up a configuration where the VM designated for training jobs doesn't run real jobs and hence doesn't get overloaded. To start, we'll create a fake tool that we'll use in our configuration.

Reviewer: ```suggestion
To demonstrate a real-life scenario and TPV's role in it, let's plan on setting up a configuration where the VM designated for training jobs doesn't run real jobs and hence doesn't get overloaded. To start, we'll create a "testing" tool that we'll use in our configuration. This testing tool can run quickly, and without overloading our small machines.
```

>    ```diff
>    --- a/group_vars/galaxyservers.yml
>    +++ b/group_vars/galaxyservers.yml

> >    @@ -123,6 +123,9 @@ galaxy_config_templates:

Reviewer: feel free to not worry about the diffs, we can resolve that at the end.

>    -            runner: local_runner
>    -            singularity_enabled: true
>    -            env:
>    -            # Ensuring a consistent collation environment is good for reproducibility.
>    -            - name: LC_ALL
>    -              value: C
>    -            # The cache directory holds the docker containers that get converted
>    -            - name: SINGULARITY_CACHEDIR
>    -              value: /tmp/singularity
>    -            # Singularity uses a temporary directory to build the squashfs filesystem
>    -            - name: SINGULARITY_TMPDIR
>    -              value: /tmp
>    -          slurm:
>    -            runner: slurm
>    -            singularity_enabled: true
>    -            env:
>    -            - name: LC_ALL
>    -              value: C
>    -            - name: SINGULARITY_CACHEDIR
>    -              value: /tmp/singularity
>    -            - name: SINGULARITY_TMPDIR
>    -              value: /tmp
>    +          tpv_dispatcher:
>    +            runner: dynamic
>    +            type: python
>    +            function: map_tool_to_destination
>    +            rules_module: tpv.rules
>    +            tpv_config_files:

> >    +              - config/tpv_rules_local.yml

Reviewer 3: tpv could only find this when I listed it with the absolute path:
`- "{{ galaxy_config_dir }}/tpv_rules_local.yml"

>    +    cores: 2
>    +    mem: cores * 4
>    +
>    +destinations:
>    +  local_env:
>    +    runner: local_runner
>    +    max_accepted_cores: 1
>    +    params:
>    +      tmp_dir: true
>    +  singularity:
>    +    runner: local_runner
>    +    max_accepted_cores: 1
>    +    params:
>    +      singularity_enabled: true
>    +    env:
>    +      # Ensuring a consistent collation environment is good for reproducibility.
>    +      LC_ALL: C
>    +      # The cache directory holds the docker containers that get converted
>    +      SINGULARITY_CACHEDIR: /tmp/singularity
>    +      # Singularity uses a temporary directory to build the squashfs filesystem
>    +      SINGULARITY_TMPDIR: /tmp
>    +  slurm:
>    +    inherits: singularity
>    +    runner: slurm
>    +    max_accepted_cores: 16
>    +    params:
>    +      native_specification: --nodes=1 --ntasks=1 --cpus-per-task={cores}
>    +
>    {% endraw %}
>    ```
>    {: data-commit="Add TPV local rules"}
>
>    In this TPV config, we have specified that the testing tool should use `2` cores. Memory has been defined as an expression and should be 4 times as much as cores, which, in this case, is 8GB.

> >    Note that the tool id is matched via a regular expression against the full tool id. For example, a full tool id for hisat may look like: `toolshed.g2.bx.psu.edu/repos/iuc/hisat2/hisat2/2.1.0+galaxy7`

Reviewer 2: Can only `hisat2` be specified here as well? In which case any version of it would be matched? If so, would be nice to explicitly clarify this.

>    This enables complex matching, including matching against specific versions of tools.
>

> >    Destinations must also be defined in TPV itself (Destinations defined in job_conf.yml are ignored by TPV). Therefore, we have moved all destinations from job_conf to TPV. In addition, we have removed some

Reviewer 2: ```suggestion
>    Destinations must also be defined in TPV itself. Importantly, note that any destinations defined in job_conf.yml are ignored by TPV. Therefore, we have moved all destinations from job_conf to TPV. In addition, we have removed some
```

> Nevertheless, you may still wish to execute these tools with memory adjusted to suit your cluster's capabilities. 
>
> 2. Edit your `files/galaxy/config/tpv_rules_local.yml` and make the following changes.
>
>    {% raw %}
>    ```diff
>    --- a/files/galaxy/config/tpv_rules_local.yml
>    +++ b/files/galaxy/config/tpv_rules_local.yml
>    @@ -30,7 +30,10 @@ destinations:
>       slurm:
>         inherits: singularity
>         runner: slurm
>    -    max_accepted_cores: 16
>    +    max_accepted_cores: 24
>    +    max_accepted_mem: 256
>    +    max_cores: 16
>    +    max_mem: 128
>         params:
>           native_specification: --nodes=1 --ntasks=1 --cpus-per-task={cores}
>
>    {% endraw %}
>    ```
>    {: data-commit="TPV clamp max cores and mem"}
>
> These changes indicate that the destination will accept jobs that are up to `max_accepted_cores: 24` and `max_accepted_mem: 256`. However, once accepted, it will forcibly clamp it down to 16 and 128 at most using the `max_cores` and `max_mem` clauses.

> > If the tool does not exceed the specified limit, it will run happily. If not, it will be clamped down to the desired range. This allows even the largest resource requirement in the shared database to be accomodated.

Reviewer 2: Probably a good idea to explicitly state what happens to any tool that defines more than `max_accepted_mem/cores`. Will it just be rejected?

Author: Yes. I'll add that clarification.

>    > {: data-cmd="true"}
>    {: .code-in}
>

### TPV reference documentation

> TPV has dedicated documentation at: https://total-perspective-vortex.readthedocs.io/en/latest/

Reviewer: ```suggestion
Please see [TPV's dedicated documentation](https://total-perspective-vortex.readthedocs.io/en/latest/) for more information.
```

# Configuring the TPV shared database

The Galaxy Project maintains a [shared database of TPV rules](https://github.com/galaxyproject/tpv-shared-database) so that admins do not have to independently rediscover ideal resource allocations for specific tools. These rules are based
on settings that have worked well in the usegalaxy.* federation. The rule file can simply be imported directly, with local overrides applied on top.

> <hands-on-title>Adding the TPV shared database to job_conf</hands-on-title>
>
> 1. Edit `group_vars/galaxyservers.yml` and add the location of the TPV shared rule file to the `tpv_dispatcher` destination.
>
>    {% raw %}
>    ```diff
>    --- a/group_vars/galaxyservers.yml
>    +++ b/group_vars/galaxyservers.yml
>    @@ -40,6 +40,7 @@ galaxy_config:
>                 function: map_tool_to_destination
>                 rules_module: tpv.rules
>                 tpv_config_files:

> >    +              - https://raw.githubusercontent.com/galaxyproject/tpv-shared-database/main/tools.yml

Reviewer: how often does this get pulled? is it cached?

Author: Pulled once on each Galaxy startup.

>    {% endraw %}
>    ```
>    {: data-commit="Importing TPV shared database via job conf"}
>
> Note how TPV allows the file to be imported directly via its http url. As many local and remote rule files as necessary can be combined, with rule files specified later overriding
> any previously specified rule files. The TPV shared database does not define destinations, only cores and mem settings, as well as any required environment vars.
> Take a look at the shared database of rules and note that some tools have very large recommended memory settings, which may or may not be available within your local cluster.
> Nevertheless, you may still wish to execute these tools with memory adjusted to suit your cluster's capabilities. 
>
> 2. Edit your `files/galaxy/config/tpv_rules_local.yml` and make the following changes.
>
>    {% raw %}
>    ```diff
>    --- a/files/galaxy/config/tpv_rules_local.yml
>    +++ b/files/galaxy/config/tpv_rules_local.yml
>    @@ -30,7 +30,10 @@ destinations:
>       slurm:
>         inherits: singularity
>         runner: slurm
>    -    max_accepted_cores: 16
>    +    max_accepted_cores: 24
>    +    max_accepted_mem: 256
>    +    max_cores: 16

> >    +    max_mem: 128

Reviewer: note that training VMs have like... 2 core and 8gb ram max usually. Should we decrease these numbers (assuming they're in GB?) (or is it MB?) to match the training VMs then, and then provide additional examples of "production" numbers in a tip box or so as "generally a good set of numbers"?

Author: Sounds like a good idea to reduce to 2 and 8.

>         params:
>           native_specification: --nodes=1 --ntasks=1 --cpus-per-task={cores}
>
>    {% endraw %}
>    ```
>    {: data-commit="TPV clamp max cores and mem"}
>
> These changes indicate that the destination will accept jobs that are up to `max_accepted_cores: 24` and `max_accepted_mem: 256`. If the tool requests resources that exceed these limits, the tool will be rejected

> > by the destination. However, once accepted, the resources will be forcibly clamped down to 16 and 128 at most because of the `max_cores` and `max_mem` clauses. Therefore, a trick that can be used here to support

Reviewer: ```suggestion
> by the destination. However, once accepted, the resources will be forcibly clamped down to 16 and 128 at most because of the `max_cores` and `max_mem` clauses. (E.g. a tool requesting 24 cores would only be submitted with 16 cores at maximum.) Therefore, a trick that can be used here to support
```

>    > {: data-cmd="true"}
>    {: .code-in}
>

# Basic access controls

You may wish to apply some basic restrictions on which users are allowed to run specific tools. TPV accomodates user and role specific rules. In addition, TPV supports tagging of tools, users, roles and destinations. These tags
can be matched up so that only desired combinations are compatible with each other. While these mechanisms are detailed in the TPV documentation, we will choose a different problem that highlights some other capabilities
- restricting a tool so that only an admin can execute that tool.

> <hands-on-title>Using conditionals to restrict a tool to admins only</hands-on-title>
>
> 2. Edit your `files/galaxy/config/tpv_rules_local.yml` and add the following rule.
>
>    {% raw %}
>    ```diff
>    --- a/files/galaxy/config/tpv_rules_local.yml
>    +++ b/files/galaxy/config/tpv_rules_local.yml
>    @@ -8,6 +8,15 @@ tools:
>         mem: cores * 4
>       testing:
>         cores: 2
>    +    rules:
>    +      - id: admin_only_testing_tool
>    +        if: |

> >    +          import os

Reviewer 3: unused import

>    +          # Only allow the tool to be executed if the user is an admin
>    +          admin_users = app.config.get( "admin_users", "" ).split( "," )
>    +          # last line in block must evaluate to a value - which determines whether the TPV if conditional matches or not

> >    +          user_email not in admin_users

Reviewer 3: ```suggestion
>    +          not user or user.email not in admin_users
```

can be matched up so that only desired combinations are compatible with each other. While these mechanisms are detailed in the TPV documentation, we will choose a different problem that highlights some other capabilities
- restricting a tool so that only an admin can execute that tool.

> <hands-on-title>Using conditionals to restrict a tool to admins only</hands-on-title>
>
> 2. Edit your `files/galaxy/config/tpv_rules_local.yml` and add the following rule.
>
>    {% raw %}
>    ```diff
>    --- a/files/galaxy/config/tpv_rules_local.yml
>    +++ b/files/galaxy/config/tpv_rules_local.yml
>    @@ -8,6 +8,14 @@ tools:
>         mem: cores * 4
>       testing:
>         cores: 2
>    +    rules:
>    +      - id: admin_only_testing_tool
>    +        if: |
>    +          # Only allow the tool to be executed if the user is an admin
>    +          admin_users = app.config.get( "admin_users", "" ).split( "," )
>    +          # last line in block must evaluate to a value - which determines whether the TPV if conditional matches or not
>    +          not user or user.email not in admin_users

> >    +        fail: Unauthorized. Only admins can execute this tool.

Reviewer: Everything about TPV looks like an improvement! :)

>    +    rules:
>    +      - id: admin_only_testing_tool
>    +        if: |
>    +          # Only allow the tool to be executed if the user is an admin
>    +          admin_users = app.config.get( "admin_users", "" ).split( "," )
>    +          # last line in block must evaluate to a value - which determines whether the TPV if conditional matches or not
>    +          not user or user.email not in admin_users
>    +        fail: Unauthorized. Only admins can execute this tool.
>
>     destinations:
>       local_env:
>    {% endraw %}
>    ```
>    {: data-commit="TPV admin only tool"}
>
{: .hands_on}

> ```bash
> 3.sh
> ```
> {: data-test="true"}

> {: .hidden}

Reviewer: is there a reason to delete this test?

Author: Thanks for the catch. That should be restored.

## Configuring TPV to process resource parameters

Lastly, we need to write a rule in TPV that will read the value of the job resource parameter form fields and decide how to submit the job.

> <hands-on-title>Processing job resource parameters in TPV</hands-on-title>
>
> 1. Create and edit `files/galaxy/dynamic_job_rules/map_resources.py`. Create it with the following contents:
>
>    {% raw %}
>    ```diff
>    --- a/files/galaxy/config/tpv_rules_local.yml
>    +++ b/files/galaxy/config/tpv_rules_local.yml
>    @@ -6,6 +6,8 @@ tools:
>         abstract: true
>         cores: 1
>         mem: cores * 4
>    +    params:
>    +      walltime: 8
>       testing:
>         cores: 2
>         rules:
>    @@ -17,6 +19,13 @@ tools:
>               # last line in block must evaluate to a value - which determines whether the TPV if conditional matches or not
>               user_email not in admin_users
>             fail: Unauthorized. Only admins can execute this tool.
>    +      - id: resource_params_defined
>    +        if: |
>    +          param_dict = job.get_param_values(app)
>    +          param_dict.get('__job_resource', {}).get('__job_resource__select') == 'yes'

> >    +        cores: int(job.get_param_values(app)['__job_resource']['cores'])

Reviewer 3: ```suggestion
>    +        cores: "{int(job.get_param_values(app)['__job_resource']['cores'])}"
```

Reviewer 3: The original line is right and does not need to change (my suggestion would change it to something that is not an integer)

>    +        params:

> >    +           walltime: int(job.get_param_values(app)['__job_resource']['time'])

Reviewer 3: ```suggestion
>    +           walltime: "{int(job.get_param_values(app)['__job_resource']['time'])}"
```

Reviewer 3: before wrapping this in "{}" the value of walltime is saved as the string: "int(job.get_param_values(app)['__job_resource']['time'])"

> ```
{: .code-out}

{% snippet topics/admin/faqs/missed-something.md step=9 %}

# More on TPV

The goal of this tutorial is to provide a quick overview of some of the basic capabilities of TPV. However, there are numerous features that we have not convered such as:
a. Custom code blocks - execute arbitrary python code blocks, access additional context variables etc.
a. User and Role Handling - Add scheduling constraints based on the user's email or role
b. Metascheduling support - Perform advanced querying and filtering prior to choosing an appropriate destination
c. Job resubmissions - Resubmit a job if it fails for some reason

> d. Linting, formatting and dry-run - Automatically format tpv rule files, catch potential syntax errors and perform a dry-run to check where a tool would get scheduled.

Reviewer: I worry about the lack of linting, should we include this somehow? It might reduce user error especially for 'complex' new syntax like python in yaml. Should we include something like https://github.com/usegalaxy-eu/ansible-tpv-lint as part of the playbook? Maybe that would be 'easy'? @mira-miracoli any thoughts on this?

Author: The ansible tpv lint would certainly be a nice thing to incorporate.
